src/h5web/dataset-visualizer/VisProvider.tsx

```diff
@@ -32,19 +31,29 @@ function VisProvider(props: Props): ReactElement {
   const rawValues = useValue(dataset.id);
   const rawDims = isSimpleShape(dataset.shape) ? dataset.shape.dims : [];
 
-  const xDim = mapperState ? mapperState.indexOf('x') : -1;
-
   const mappedValues = useMemo(() => {
     if (!rawValues) {
       return undefined;
     }
 
-    if (rawDims.length === 2 && xDim === 0) {
-      return transpose(rawValues);
+    if (rawDims.length !== 0) {
+      return ndarray(rawValues.flat(), rawDims);
     }
 
     return rawValues;
-  }, [rawValues, rawDims.length, xDim]);
+  }, [rawValues, rawDims]);
+
+  const values = useMemo(() => {
+    if (mapperState === undefined) {
+      return mappedValues;
+    }
+    const slicingIndices = mapperState.map(val => (isNumber(val) ? val : null));
+    const slicedDims = rawDims.filter((d, i) => !isNumber(mapperState[i]));
+    return ndarray(
+      unpack(mappedValues.pick(...slicingIndices)).flat(),
+      slicedDims
+    );
+  }, [mappedValues, mapperState, rawDims]);
```

Comment:
Slicing is now done directly in the `VisProvider` so that it provides the sliced values. The `useMemo`s could probably be merged.